### DISCUSSION TITLE
[Feature Request] Include vertex hit info in input info.

### DISCUSSION THREAD

**Discussion by DanielWillett** [SmartlyDressedGames/Unturned-3.x-Community#4577] [Ideas]
I'm proposing adding `RaycastHit.triangleIndex` and `RaycastHit.barycentricCoordinate` to the replicated data in InputInfo when you hit or shoot an object other than a player/npc like placeables, objects, resources, vehicles, etc.

This will allow us to add special behavior when a specific part of an object is shot or punched, along with the already existing `colliderTransform`. It would be trivial for us to get any vertex data from the mesh collider with this info, like UVs, vertex colors, etc.

Example usages: 
* damage multipliers to different parts of vehicles
  * this is the main one since moving vehicles plus random dude with 651ms of ping makes it impossible to raycast server-side
* highly precise screens implemented using tessellated planes
* spawning custom hit effects depending on what was hit.

It would be fairly easy and cheap to add and would allow plugins to do a lot of cool things with dynamic hit actions.

**Comment by Johnaadms** [comment]
What are RaycastHit.triangleIndex and RaycastHit.barycentricCoordinate?

RaycastHit.triangleIndex: Represents the index of the triangle that was hit on the mesh collider of the object.
RaycastHit.barycentricCoordinate: Provides the barycentric coordinates of the hit point within the triangle, indicating how the hit point is interpolated between its vertices.
How would adding these to InputInfo benefit the system?
[priceinfo](https://priceinfo.org/)
Enhanced Behavior Differentiation: Allows for specific actions or effects based on where exactly an object was hit. For instance, different damage multipliers on various parts of vehicles or triggering custom hit effects based on hit locations.
Precision in Interaction: Provides precise data about the hit location, which can be crucial for implementing complex behaviors like highly accurate screen interactions or detailed hit effects.

**Reply by Jdance-Media** [reply]
Huh, almost as if that is exactly what Blaze said but just formatted a tad differently.

Anyways, raycasts server-side are incredibly slow. How do I know this? I tried generating a chart on the server. So any measure to limit how many raycasts we need to do is a good measure in my book.

**Comment by GazziFX** [comment]
Isn't it can be retrieved from local position?

**Reply by DanielWillett** [reply]
Not for vehicles as they could've moved by the time the server gets the hit.

**Reply by GazziFX** [reply]
> Not for vehicles as they could've moved by the time the server gets the hit.

I mean local position property can be added and it will be the same regardless of vehicle position
```csharp
raycastInfo.localPosition = raycastInfo.transform.InverseTransformPoint(raycastInfo.point);
```
The only drawback is it will be hard to determine position on a limbs for rigged entities like player or animal as they have separate moving parts with colliders

**Reply by DanielWillett** [reply]
I suppose that would work too, just figured triangleIndex and barycentricCoordinate would be basically free since it wouldn't require any extra math or anything, also wouldn't suffer from precision issues, perhaps both could be added, although as you said it wouldnt work as well on skeletal meshes or turrets that can rotate.

**Comment by SDGNelson** [comment]
I think I have some good news for you! When I implemented the NetId-relative transform replication, `InputInfo.colliderTransform` became replicated from client to server. If you give the item you want to detect its own child transform with a collier it should be doable!

**Reply by DanielWillett** [reply]
Yeah, that's how we're doing it now but were hoping to switch to vertex colors (we're doing variable vehicle damage based on where it's hit).
